feat: Allow reading part of a file

HarelM opened the pull request:
## PR Description

Original PR added a new `readRange` method.

Has since been refactored by @OS-pedrogustavobilro to Implement partial read of files by adding optional `offset` and `length` to `readFile` and `readFileInChunks`.

## Context

Allow reading only part of a file

- See https://github.com/ionic-team/capacitor-filesystem/issues/19

## Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply -->
- [ ] Fix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Refactor (cosmetic changes)
- [ ] Breaking change (change that would cause existing functionality to not work as expected)


## Tests

Refer to unit tests for validating changes.

For the team reviewing this PR, you can [test using this OutSystems app](https://intranet.outsystems.net/MABS_Lite/BuildDetail?id=ae45fffaaa75ec2c3545e832c2fb75c55294eacf&stg=dev) -> Click "Files", write a file, then click it to open a new screen to read, there you'll be able to expand options for offset and length.

HarelM committed:
feat: Add readRange method

HarelM committed:
Update changelog

OS-pedrogustavobilro commented:
Hey @HarelM, from a quick look this PR looks good, but I do believe a different approach would be better - instead of adding that feature in a separate method, add it to the existing methods `readFile ` and `readFileInChunks`, basically adding the offset and length fields as optional fields in `IONFILEReadOptions` - so that they can also be called from the existing methods in the Capacitor plugin.

Unless you added them in separate methods for a particular reason?

HarelM commented:
Well, there's no strong argument towards adding a different call, but I didn't want to complicate the existing implementation with the code that I've added. 
I'm guessing that read file and read file in chunks can also be overloaded to the same method, but splitting them makes things easier in terms of separation of concerns.
There's also the input stream vs random access file.
If you think it's a better approach I can try implementing that, but I'm more concerned about what I'll break when doing that, but I don't mind trying.

OS-pedrogustavobilro commented:
> If you think it's a better approach I can try implementing that, but I'm more concerned about what I'll break when doing that, but I don't mind trying.

I'll be taking a closer look at both approaches (separate method vs overloading), and may do edits to your PRs if I find other approach to be better, saves you some effort 😅 if you don't mind.

HarelM commented:
That would be great, yes, feel free to change anything you want.

OS-pedrogustavobilro committed:
refactor: Use offset and length in existing file methods

An alternative to creating a readRange method. This way it allows us to better use the functionality in consumers of the native library (e.g. capacitor plugin)

OS-pedrogustavobilro committed:
refactor: Move checks of offset and length to controller

OS-pedrogustavobilro committed:
docs: Update README and CHANGELOG

OS-pedrogustavobilro commented:
Hey @HarelM, I ended doing a near-complete re-write of the code. It's not that your code was bad by any means, but upon inspection of the feature I felt it would be proper to add this to exiting methods `readFile` and `readFileInChunks` (I imagine use cases for the latter aren't that many, but I could think of like trying to process a file in chunks and the process being paused and wanting to resume some time after). Plus, using `RandomAccessFile` would work fine for a local file where you have a proper file path to it, but wouldn't work with `content://` URIs

So I ended up including this partial read of file in the existing methods with `InputStream`, and from what I tested so far everything is working fine.

The main concern I had was if `InputStream#skip` would be less efficient than `RandomAccessFile#seek`, but from testing in different android devices with large file size offsets (100MB to 1GB), they seemed to have taken the same amount of time (both under 1 millisecond), so all's good there. Perhaps Android uses `seek` under the hood in `skip`.

Anywho, I'm still gonna run some more tests in internal apps we have, but I think this PR is in a near-merge state now - will request a review from a colleague in the coming days and move to your iOS PR (which, for similar reasons, I may need to do some re-write to support more use cases for the library / capacitor plugin that consumes the library).

Let me know if you have any questions. Thanks!

HarelM commented:
As long as this library has this feature I'm happy 😀
I want to replace the cordova file system plugin.

OS-pedrogustavobilro committed:
refactor: Rename length default value

OS-pedrogustavobilro committed:
chore(release): Prepare version 1.1.0
